src/cpp/core/FileSerializer.cpp

```diff
@@ -24,13 +24,62 @@
 #include <boost/algorithm/string/trim.hpp>
 #include <boost/algorithm/string/predicate.hpp>
 #include <boost/iostreams/copy.hpp>
+#include <boost/thread.hpp>
 
 #include <shared_core/FilePath.hpp>
+#include <core/DateTime.hpp>
 #include <core/StringUtils.hpp>
 
 namespace rstudio {
 namespace core {
 
+namespace {
+
+bool isFileLockedError(const Error& error)
+{
+   // exclusive file access is only present on Windows
+#ifndef _WIN32
+   return false;
+#endif
+   return (error && error.getCode() == ERROR_SHARING_VIOLATION);
+}
+
+Error openFileForWritingWithRetry(const FilePath& filePath,
+                                  bool truncate,
+                                  int maxOpenRetrySeconds,
+                                  std::shared_ptr<std::ostream>* pOfs)
+{
+   using namespace boost::posix_time;
+
+   ptime startTime = second_clock::universal_time();
+   Error lastError;
+
+   // do not allow negative values - regular signed int was chosen here for
+   // easier integration with other parts of the codebase
+   if (maxOpenRetrySeconds < 0)
+      maxOpenRetrySeconds = 0;
+
+   while (true)
+   {
+      lastError = filePath.openForWrite(*pOfs, truncate);
+
+      // if the error is a non file lock error, then we should just return it
+      if (!isFileLockedError(lastError))
+         return lastError;
+
+      // stop retrying if we've spent more than the requested amount of time
+      if ((second_clock::universal_time() - startTime) >= seconds(maxOpenRetrySeconds))
+         break;
+
+      // wait a moment before retrying
+      boost::this_thread::sleep(milliseconds(500));
```

Comment:
Does this actually pause the whole RPC thread? I'm a little concerned that if (a) we auto backup changes every 500ms (the default), and (b) every (other) attempt to write backups pauses the RPC thread for 500+ms as we wait for GDrive (etc) to release the lock, the RPC thread is going to spend a pathologically large time locked.

Reply:
It does since it is sleeping. I did my testing with auto save on 500 ms and the performance seemed good. The vast majority of saves are non conflicting - I'd say one in every 15-20 saves produced a stall waiting for the file to become available in my testing. I didn't notice any ill effects from a UX perspective.

This also reduces a bunch of complexity as opposed to retrying the operations in the front end and having to re-queue operations.

Reply:
Was your testing on a system with a fast SSD and internet connection? I think that GDrive and friends can have an exclusive lock for much longer on some environments, leading to more save conflicts. 

I think we need to give up faster on autosaves. The user generally has another chance to save the file (and will get warned if they try to exit with changes that have never been backed up); it feels wrong that there's a possibility of blocking the entire RPC thread for 15 seconds for an operation the user didn't initiate. 

15s isn't unreasonable for a change that must succeed, so I'd suggest a second pref for autosave (auto_save_retry_timeout?) as both a shorter fuse and an escape hatch; it could give up after a second or two, and setting it to 0 would result in a best-effort autosave that just attempts to write the file if it can (the front end could read this value and avoid popping up "file is in use" errors).

Reply:
What about not retrying for autosaves period? The whole point of autosaves is that it saves continuously and periodically. Since most operations are successful, we could simply ignore this particular failure for autosaves entirely, with the knowledge that future autosaves will almost assuredly work (since the exclusive file lock will be given up as soon as syncing is done).

We can keep everything as is currently for manual saves, since they are a user initiated action.